p2p/protocol/head/head.go

```diff
@@ -0,0 +1,99 @@
+package head
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+)
+
+const closeTimeout = 30 * time.Second
+
+var log = logging.Logger("go-legs-head")
+
+func deriveProtocolID(topic string) protocol.ID {
+	return protocol.ID("/legs/head/" + topic + "/0.0.1")
```

Comment:
If `0.0.1` represents the version would it make sense to move that after `/legs/`?

Reply:
I don't think anything technically prohibits the version being in a certain place (or even not being there at all). There is a [helper](https://github.com/libp2p/go-libp2p-core/blob/master/helpers/match.go#L16) that expects the version to be in the last bit, but that's only if you use `SetStreamHandlerMatch`. In this case we don't, we use the simpler `SetStreamHandler`.

And even if we did use `SetStreamHandlerMatch` we could write our own matcher that got the version from wherever.

But I'm not convinced (either way) what is the correct thing here. My intuition is that the protocol id should serve to quickly refine if another protocol id is compatible and each part expands on the previous part (this is why the version isn't first). The sooner we can figure out if we are a compatible protocol the better.

In this case if the topic is different, we know this isn't a compatible protocol (regardless if the version matches). If the version were sooner, we would be version matching, then realize the topic is different at the end (possibly wasting some compute in version matching when we would have know the topic was incompatible anyways).

What's a little weird about this I think, and why it feels like it would make sense to move the version to before the topic, is that the topic string is part of the protocol id. This is a little weird in that a topic itself doesn't change the protocol at all. It's used as part of the protocol ID so that we have an unique stream per topic. It would almost make sense to think of this as the stream id instead.